### PR TITLE
report EPIPE and EOF errors betterer

### PR DESCRIPTION
**Pull request by shiftkey**
There's a couple of common-but-obscure errors sitting in Haystack that I'd like to understand more about. 

This one only occurs on macOS:

```
error: write EPIPE
exports._errnoException (util.js:1050:11)
WriteWrap.afterWrite (net.js:813:14)
```

This one only occurs on Windows:

```
error: write EOF
exports._errnoException (util.js:1050:11)
WriteWrap.afterWrite (net.js:813:14)
```

Digging into more details about these errors, I stumble upon this project: https://github.com/mhart/epipebomb

> By default, node throws `EPIPE` errors if `process.stdout` is being written to and a user runs it through a pipe that gets closed while the process is still outputting (eg, the simple case of piping a node app through `head`).
> 
> This seemed a little overzealous to me, so I wrote this to suppress such errors.

I'm not quite sure how this might apply to Desktop, but the details of this suppression was very interesting: they do this by attaching to the `error` event on `process.stdout`.

This is interesting, for a couple of reasons:

 - #3954 was a situation where I did the exact same thing to listen for `error` events from the spawned process (to stop the renderer from crashing)
 - the stack trace for both errors are the same, which makes me think I can handle both errors by handling the `error` event on `process.stdout`

~~There's also `GitProcess` in `dugite` that might be responsible for these errors, and I'll address that separately to this, but~~ this PR should help us identify the source of these errors better than what's currently delivered to Haystack.

### CODE REVIEW

**Comment by shiftkey** [comment]
> There's also GitProcess in dugite that might be responsible for these errors, and I'll address that separately to this, 

Actually, I'm going to leave that aside for now as I think the `execFile` callback in `dugite` (which handles ENOENT and should bubble up other error codes when we `await` it) should cover us: https://github.com/desktop/dugite/blob/master/lib/git-process.ts#L183

The fact that the error lacks any Desktop source code in the stack trace makes me think this is related to the spawn-based APIs, and we need a similar fix to what we did in #3954.

**Comment by shiftkey** [comment]
@desktop/core I'd like to get this into 1.1 as it's been around for the last couple of versions, and it might help ameliorate the error reports we're seeing. Let me know if you have any feels on this.

**Comment by nerdneha** [comment]
i'm 👍 for some more info, will let others chime in

**Comment by niik** [comment]
I don't believe that stdout is the problem here and I'd be surprised if the logging in this PR picks up anything useful for us (though I have a strong track record of being wrong). To me this looks like an error that I'd expect if I'm piping data to a child process which closes its *stdin*.

Here's a small repro

```js
require('child_process')
  .spawn("head", ["-c", "1"])
  .stdin.end('x'.repeat(1024 * 1024))
```

output

```
process-input-closed markus$ node index.js
events.js:163
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at exports._errnoException (util.js:1050:11)
    at WriteWrap.afterWrite [as oncomplete] (net.js:813:14)
```

The only case that I see stdout writes being the cause of this error is for data that we ourselves (i.e. desktop) are writing to stdout which should only be log messages (for the main app) and since we haven't made any real changes to that in the release where these errors spiked I think we should consider that to be of fairly low probability.

We did, however, added two new git calls that write to git's standard input in the last release, `mergeTrailers` and `parseTrailers`. I strongly suspect that it's one of those two that's the cause, that for some reason `git interpret-trailers` is terminating abruptly but we never see that actual error because node throws the stdin error first and we terminate the app before we can ever see the actual error from git and the exit code.

As far as I'm concerned a child process closing its stdin should not warrant an error. I think we should suppress it and wait for dugite's error logic to pick up on the exit code and parse (if any) stderr/stdout.

This won't solve the underlying problems obviously but our errors should now be thrown in a way that we can handle them and hopefully catch some stacks that allow us to pinpoint which git call is prematurely closing its input and why.

So to summarize I think we should address this in dugite by suppressing the error both for `execFile` and `spawn`. For `spawn` we could conceivably do this in the app ourselves without involving dugite but for `execFile` we can't since the `processCallback` where we'd be able to attach our error handler runs after we flush the stdin data.

Should consumers of dugite want to be notified when stdin closes they can either subscribe to `ChildProcess.stdin.on('close')` or `ChildProcess.stdin.on('error')` and override our suppression but I think this would make for a better default behavior.

I'll open a PR against dugite and we can continue the discussion there.

**Comment by shiftkey** [comment]
> I'll open a PR against dugite and we can continue the discussion there.

@niik I'm into it

**Comment by shiftkey** [comment]
> I don't believe that stdout is the problem here and I'd be surprised if the logging in this PR picks up anything useful for us (though I have a strong track record of being wrong). 

Thinking about this some more, the `write EOF` reports in Haystack (like #2041) have around for a while and seem to be related to committing a large numbers of files, and one of the commands we depend on - `update-index` does use `stdin` to pass these file paths to Git:

https://github.com/desktop/desktop/blob/bcf0b669e29ec21131ed9cd8a4613caac1a01c06/app/src/lib/git/update-index.ts#L90-L94

I'm still annoyed to have not been able to reproduce this behaviour myself, but I'm keen to see what you dig up. I'm also happy to pick this up and get a fix into `dugite` to improve our diagnostics, if you've got enough on your plate next week.
